### DISCUSSION TITLE
Share XSatate Machine between various Vue Components using '@xstate/vue'

### DISCUSSION THREAD

**Discussion by santicros** [statelyai/xstate#1479] [General] [answered]
Hey there!!!

I was wondering is there any in which we can share a state machine between different components using the @xstate/vue package?
I've seen there is a useService hook, but can't see how should I use it in different components. I want to access the same state and context in all components, like I would do with Vuex.
I've seen in React they use the React.Context to share the machine, but how should it be done in Vue? Has anyone done it?

Thanks so much!

**Comment by davidkpiano** [comment, marked as answer]
Hey @santicros -

You can define the service outside of your components:

```js
import { interpret } from 'xstate';

export const someService = interpret(someMachine).start();
```

And then you can import that `someService` and use it via `useService(...)`.

Does that help?

**Reply by santicros** [reply]
Hello @davidkpiano! Thanks for the fast response. That was really what I was searching 😍.

I've tried and exported a custom hook for that machine that can be reused on any component.

If it can help anyone a more detailed explanation:
```js
import { createMachine, interpret } from "xstate";
import { useActor } from "@xstate/vue";

// I create my XState machine
const toggleMachine = createMachine({
  id: "toggle",
  initial: "inactive",
  states: {
    inactive: {
      on: { TOGGLE: "active" },
    },
    active: {
      on: { TOGGLE: "inactive" },
    },
  },
});

// Create a service as @davidkpiano commented above
const service = interpret(toggleMachine).start();

// Create a custom service hook for this machine, that takes the service previously created. This way you don't have to import on every component the useService hook and the service created.
export const useToggleMachine = () => {
  return useActor(service);
};

```

Then in any Vue component:
```vue
<template>
  <div>
    <p>{{ state.value }}</p>
  </div>
</template>

<script>
import { useToggleMachine } from "../store/useToggleMachine.js";
export default {
  setup() {
    const { state, send } = useToggleMachine();
    return {
      state,
      send,
    };
  },
};
</script>

```

**Reply by iksaku** [reply]
Thanks @davidkpiano for the solution! Wasted basically 0 time googling around.

Just as a note... For anybody looking at this one (or more) years later, instead of using `useService`, use the `useActor` hook 😉

**Reply by santicros** [reply]
> Thanks @davidkpiano for the solution! Wasted basically 0 time googling around.
> 
> Just as a note... For anybody looking at this one (or more) years later, instead of using `useService`, use the `useActor` hook 😉

Thanks for the response @iksaku! I've updated my demo with `useActor` and `createMachine` ;)

**Comment by yaffol** [comment]
Hi,

Is there a way to implement this within Vue 2, or does it depend on Vue 3?

I would like to be able to share a machine between Vue components (and some plain Javascript) but an tied to Vue 2 given a dependency on Vuetify.

Thanks for a great library!

**Reply by Andarist** [reply]
Your question is quite vague - could you rephrase so we would know exactly what "implement *this*" means in this context? Could you give an example of what you are trying to achieve and what you have tried so far?

**Reply by santicros** [reply]
With Vue 2 you can't use the version of `@xstate/vue 0.4.0` or next ones as they will be only for Vue 3. However, you can keep using the 0.3.0 version together with the composition api:

``` bash
npm i @xstate/vue@0.3.0 @vue/composition-api
```

**Comment by yaffol** [comment]
Thank you for the information, and apologies for the lack of specifics in my initial question.

I have a state machine which is defined in its own file, I am importing that into a Vue2 component, and interpretting and starting the machine within that component (in the `data()` and `created()` hook respectively).

I also have another library which is imported into the component (a simple authentication handling library), and I'd like to be able to send messages to the machine, and get the current state and context from both within the Vue2 component, and within that external library.

My impression was that the means to support this via `useService()` was only possible with the Vue composition API, and hence Vue 3 or the back-ported composition api package. So I think it will work by using `@xstate/vue@0.3.0 @vue/composition-api` as @santicros suggests.

I have followed the suggestion of installing `npm i @xstate/vue@0.3.0 @vue/composition-api` and am glad to report it works as expected - I can interpret and start the machine within the machine's external definintion file, and expose a call to `useService(..)` on that interpreted machine.

```
export const authService = interpret(authMachine)
  .onTransition((state) => console.log(state.value))
  .start()

export const useAuthMachine = () => {
  return useService(authService)
}
```
Which I can import into either a vue component using the compistion API syntax
```
  setup () {
    const { state, send } = useAuthMachine()
    return {
      state,
      send
    }
  }
```
or into an external file
`import { authService } from '@/machines/authMachine'`

**Reply by OmarIbrahimDymk** [reply]
@yaffol may i know what @vue/composition-api version are you using?

Im getting error as shown below,
![image](https://user-images.githubusercontent.com/64179389/109114790-1612a180-7779-11eb-876b-c2f99078caad.png)

for my machine, i follow exact toggle machine written by @santicros  above.

im using,
nuxt 2.14.5
@vue/composition-api 0.5.0
@xstate/vue 0.3.0
xstate 4.16.2
